### ISSUE TITLE
[Enhancement]: QueryNode separate Search/QueryTasks logic and provide gRPC APIs

### ISSUE THREAD

**Issue by congqixia** [milvus-io/milvus#15724]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

QueryNode separate Search/QueryTasks logic and provide gRPC APIs

### Why is this needed?

To support memory replica

**Comment by congqixia** [comment]
![Design_SearchQuery_gRPC_current](https://user-images.githubusercontent.com/84113973/155452146-1e6c8c32-86b4-4807-a561-94ec8a315213.png)

**Comment by xiaofan-luan** [comment]
Questions, how do we know all the segments should be involved in a search/query?

**Comment by congqixia** [comment]
> Questions, how do we know all the segments should be involved in a search/query?

Proxy shall maintain a cache of the segment allocation.
In case of the cache is out-of-date, there are several ways to handle this problem:
1. Proxy refresh the case in a short period of time.
2. Each search/task request contains invalid segment id(which current query node does not handle) will have a quick response msg in stream. So proxy can update the cache and resend the missing segments.
3. Global sealed id list will be returned immediately after a `QueryNode` has a corresponding dml channel (of the collection/partition) in it.  This mechanism is used as a fast short-cut to detect cache invalidation.

**Comment by xiaofan-luan** [comment]
<img width="750" alt="image" src="https://user-images.githubusercontent.com/83447078/155563981-8e2340a0-4eeb-4ab9-8952-ea4d4868573a.png">

**Comment by xiaofan-luan** [comment]
I have a proposal to split the search into 3 stage:
1) proxy send request to all shards on querynodes
2)each querynodes send request to all sealed segments belongs to this shard, and on each querynode we do local reduce and merge result if Q1 contains multiple segment of a shard.
3) Each shard do reduce from all the result from all sealed segment and growing segment
4) proxy reduce all results from all shards.

**Comment by xiaofan-luan** [comment]
Based on this implementation, we can also change the delta channel to a rpc implementation. Shard is responsible to send delta log to sealed segment.

Another good point of current implementation is the consistency would be still followed, as long as delete nofication and insert is executed in order, then sealed segment do not need to receive TT but the consistency rule is not breaked.

@sunby @congqixia @xige-16 what do you guys think?

**Comment by congqixia** [comment]
> I have a proposal to split the search into 3 stage:
> 
>     1. proxy send request to all shards on querynodes
>        2)each querynodes send request to all sealed segments belongs to this shard, and on each querynode we do local reduce and merge result if Q1 contains multiple segment of a shard.
> 
>     2. Each shard do reduce from all the result from all sealed segment and growing segment
> 
>     3. proxy reduce all results from all shards.

@xiaofan-luan I'm kind of confuse of this description:
1. What's the definition of a `Shard`?
2. Assuming a `Shard` is a group of segments
   a. By "proxy send request to all shards on querynodes", does it mean broadcasting the Search/Query request?
   b. If not, does it mean send request to the leader one?

**Comment by xiaofan-luan** [comment]
> 1. What's the definition of a `Shard`?

What's the definition of a Shard? -> shard in milvus is the dml channel in pulsar, but here I'm indicating the querynode consume to this dml channel.

2. when create collection, user specify the shard numbers, by default is 2 or 4 I guess. Proxy only send request to querynode hold the dml channel, and those querynode redirect request to sealed segment holders(other query nodes)

**Comment by congqixia** [comment]
> Proxy only send request to querynode hold the dml channel, and those querynode redirect request to sealed segment holders(other query nodes)

So this QueryNode, which holds the dml channel, will behave like a leader.

**Comment by xiaofan-luan** [comment]
> > Proxy only send request to querynode hold the dml channel, and those querynode redirect request to sealed segment holders(other query nodes)
> 
> So this QueryNode, which holds the dml channel, will behave like a leader.

Yep, we name it shard holder or shard coordinator. No matter which querynode take the dml channel will be named

**Comment by congqixia** [comment]
This feature is under development. Let's keep this issue open.

**Comment by congqixia** [comment]
This issue is resolve by memory replica feature #16291, closing.
